### ISSUE TITLE
Improve the layout algorithm of data segments

### ISSUE THREAD

**Issue by TerrorJack** [tweag/asterius#289]
Currently, our data segment layout algorithm simply lays out the segments based on the symbol lexicographical order, and each segment aligns to 16 bytes. Each group of static data gets a wasm memory segment. The downsides are:

* For very large output wasm modules, the memory segment number can exceed the `100000` count limit. (`276576` when `--no-gc-sections` is on, with a simple TH splice)
* Some wasted space due to 16-byte alignment
* Read-only and read-write segments are not separated; this may be required for improved memory traps

The new layout algorithm should:

* Determine the accurate alignment requirement for each group of static data, using GHC's Cmm-related logic
* Separate read-only and read-write segments into different static blocks
* ~~(Optionally)~~ Combine adjacent segments and fill the gap with zero or trash bytes, to workaround the data segment count limit

**Comment by hamishmack** [comment]
In case there is any confusion I meant put `#289` in my commit message, but meant https://github.com/input-output-hk/haskell.nix/pull/289 (not this one).

**Comment by TerrorJack** [comment]
Even when `--no-gc-sections` is disabled, for certain TH splices in the wild we've witnessed a huge amount of segments:

```
Building library for language-c-quote-0.12.2.1..

Language/C/Smart.hs:69:21: error:
    • Exception when trying to run compile-time code:
        EvalException {evalError = "CompileError: WebAssembly.compile(): data segments count of 121512 exceeds internal limit of 100000 @+33045696"}
      Code: Language.Haskell.TH.Quote.quoteExp
              cexp "3.141592653589793238"
    • In the quasi-quotation: [cexp|3.141592653589793238|]
   |
69 |     pi            = [cexp|3.141592653589793238|]
   |
```

So improving the layout algorithm is also related to TH improvements.

**Comment by TerrorJack** [comment]
Should have been closed by #420
